### DISCUSSION TITLE
npm postinstall lifecycle hook and dependencies

### DISCUSSION THREAD

**Discussion by bgutschke** [renovatebot/renovate#13783] [Suggest an Idea]
Hello,

I am a little bit unsure if this is an issue, but I wanted to share, since it might help others as well.

We are using the self-hosted variant of Renovate and have a lot of @nrwl/nx workspaces. Nx comes with the npm lifecycle hook `"postinstall": "node ./decorate-angular-cli.js && ngcc --properties es2015 browser module main"`, which led to package-lock.json files not being updated.

This is due to the default configuration option [skipInstalls](https://docs.renovatebot.com/self-hosted-configuration/#skipinstalls) and ngcc in postinstall.

Default setting for [skipInstalls](https://docs.renovatebot.com/self-hosted-configuration/#skipinstalls) is false, which runs npm with the flag `--package-lock-only` (see https://docs.npmjs.com/cli/v8/commands/npm-install), which leads to the postinstall task to be failing, since ngcc is not being installed.

Hope this helps!

**Comment by viceice** [comment]
This is intended renovate behavior, so you need to manually allow scripts

https://docs.renovatebot.com/self-hosted-configuration/#allowscripts

**Reply by bgutschke** [reply]
Sure, this is the intended behavior, but figuring out why the npm postinstall failed - and I am not talking about postUpgradeTasks - was   quite hard to figure out. 

And just setting allowScripts to true would not have helped here, since ngcc as a dependency still would not have been installed. 

skipInstalls is the solution as far as I can see, if you use/depend on dependencies in your postinstall npm lifecycle scripts. Although that means that all dependencies get installed in this step, while at least in this case only @angular/compiler-cli aka ngcc was needed.

btw: I really <3 Renovate. Thank you very much for your work!

**Reply by rarkins** [reply]
How would you normally install only one of your devDependencies instead of the full package.json?

**Reply by bgutschke** [reply]
I would probably just use `npm install`. In this particular case `npm install -D @angular/compiler-cli@<version>`. 

Although here I would need to make sure, that the version exactly matches the version in the package.json, so I do not change the package.json file. Installing it globally with the `--global` flag might also be an option.

But for that we would need something like `preUpgradeTasks`. Is that something worth to consider?

**Reply by rarkins** [reply]
Yes, I'm ok with a feature request for `preUpgradeTasks` if you can also accompany it with a reproduction/mock config of how you'd like to use it (i.e. similar to what you wrote above).

If it were me, I'd just set `skipInstalls` to false though

**Reply by bgutschke** [reply]
Thx! I will check out the skipInstalls setting and how it impacts the Renovate runs times and might come back to this.
